core-processor/src/ext.rs

```diff
@@ -424,18 +470,38 @@ impl EnvExt for Ext {
         self.check_forbidden_call(self.context.message_context.reply_destination())?;
         self.charge_expiring_resources(&msg)?;
 
-        if delay == 0 {
-            self.charge_gas(self.context.message_context.settings().sending_fee())?;
-        } else {
-            self.charge_gas(
-                self.context
-                    .message_context
-                    .settings()
-                    .scheduled_sending_fee(),
-            )?;
+        self.charge_sending_fee(delay)?;
+
+        let result = self.context.message_context.reply_commit(msg, delay, None);
+
+        self.return_and_store_err(result)
+    }
+
+    fn reservation_reply_commit(
+        &mut self,
+        id: ReservationId,
+        msg: ReplyPacket,
+        delay: u32,
+    ) -> Result<MessageId, Self::Error> {
+        self.charge_gas_runtime(RuntimeCosts::ReservationReplyCommit)?;
+
+        self.check_forbidden_call(self.context.message_context.reply_destination())?;
+
+        self.check_message_value(msg.value())?;
+        let _gas_limit = self.check_gas_limit(msg.gas_limit())?;
+        // TODO: gasful sending (#1828)
+        self.charge_message_value(msg.value())?;
+
+        self.charge_sending_fee(delay)?;
+
+        if !self.context.gas_reserver.exists(id) {
```

Comment:
seems now you can twice use reservation id to send a message, isn't it? provide test, please